packages/hebao_v1/contracts/aux/BatchTransactor.sol

```diff
@@ -5,48 +5,21 @@ pragma experimental ABIEncoderV2;
 
 import "../lib/OwnerManagable.sol";
 import "../lib/Drainable.sol";
-import "../thirdparty/chi/IChiToken.sol";
+import "./ChiDiscount.sol";
 
 
 /// @title BatchTransactor
 /// @author Daniel Wang - <daniel@loopring.org>
-contract BatchTransactor is Drainable, OwnerManagable
+contract BatchTransactor is Drainable, ChiDiscount, OwnerManagable
 {
-    IChiToken public immutable chi;
-
-    // See:
-    // - https://github.com/1inch-exchange/1inchProtocol/blob/a7781cf9aa1cc2aaa5ccab0d54ecbae1327ca08f/contracts/OneSplitAudit.sol#L343
-    // - https://github.com/curvefi/curve-ren-adapter/blob/8c1fbc3fec41ebd79b06984d72ff6ace3198e62d/truffle/contracts/CurveExchangeAdapter.sol#L104
-    modifier discountCHI(uint maxToBurn)
-    {
-        uint gasStart = gasleft();
-        _;
-        if (maxToBurn == 0) return;
-        uint gasSpent = 21000 + gasStart - gasleft() + 16 * msg.data.length;
-        uint fullAmountToBurn = (gasSpent + 14154) / 41947;
-        uint amountToBurn = fullAmountToBurn > maxToBurn ? maxToBurn : fullAmountToBurn;
-        if (amountToBurn > 0) {
-            // Only managers are allowed to burn tokens using this contract.
-            require(isManager(msg.sender), "UNAUTHORIZED");
-            chi.freeUpTo(amountToBurn);
-        }
-    }
-
-    constructor(
-        address _chi             // Mainnet: 0x0000000000004946c0e9F43F4Dee607b0eF1fA1c
-        )
-    {
-        chi = IChiToken(_chi);
-    }
-
     function batchTransact(
         address target,
-        bytes[] calldata txs,
-        uint[]  calldata gasLimits,
-        uint    numGasTokensToBurn
+        bytes[]   calldata txs,
+        uint[]    calldata gasLimits,
+        ChiConfig calldata chiConfig
         )
         external
-        discountCHI(numGasTokensToBurn)
+        discountCHI(chiConfig)
```

Comment:
Will need to have some check like `require(isManager(msg.sender), "UNAUTHORIZED");` otherwise anyone can just burn gas tokens burnable by this contract.

`onlyManager` here could make sense here if this contract would used as some privileged contract in the called contract, otherwise `require(isManager(msg.sender), "UNAUTHORIZED");` only when gas tokens are burned will be a bit more efficient.